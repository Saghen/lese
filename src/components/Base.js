import styled from "@emotion/styled";
import { propertyGenerator } from "../helpers";

const getTextProperties = propertyGenerator([
  "color",
  "fontSize",
  "fontDecoration",
  "fontWeight",
  ["textAlign", { default: "center", property: "text-align" }]
]);

const getSizeProperties = propertyGenerator([
  [
    "height",
    ({ responsive, height }) =>
      responsive ? `max-height: ${height}; height: 100%` : `height: ${height}`
  ],
  [
    "width",
    ({ responsive, width }) =>
      responsive ? `max-width: ${width}; width: 100%` : `width: ${width}`
  ],
  "padding",
  "margin"
]);

const getLayoutProperties = propertyGenerator([
  ["block", () => "display: block"],
  ["inlineBlock", () => "display: inline-block"],
  ["relative", () => "position: relative"]
]);

const getCosmeticProperties = propertyGenerator([
  "border",
  "borderRadius",
  "background"
]);

export default styled.div`
  ${getLayoutProperties}
  ${getSizeProperties}
  ${getTextProperties}
  ${getCosmeticProperties}
`;
