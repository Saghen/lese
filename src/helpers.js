import styled, { css } from "styled-components";

export function arrayToCSS(array) {
  return `${array.join(";")};`;
}

export function getFlexProperties({
  column,
  xAlign,
  yAlign,
  wrap,
  separation
}) {
  const properties = [];
  properties.push("display: flex");

  // Set defaults
  xAlign = xAlign && returnDefault(xAlign, "string", "center");
  yAlign = yAlign && returnDefault(yAlign, "string", "center");
  wrap = wrap && returnDefault(wrap, "string", "wrap");

  if (wrap) properties.push(`flex-wrap: ${wrap}`);

  if (column) properties.push("flex-direction: column");

  if (xAlign)
    properties.push(
      column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`
    );
  if (yAlign)
    properties.push(
      !column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`
    );

  if (separation)
    properties.push(
      `
        > * {
          margin: 0;
        }
        > * + * {
          ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`}
        }
      `
    );

  arrayToCSS(properties);
}

export function getTextProperties({
  color,
  fontSize,
  fontDecoration,
  fontWeight
}) {
  const properties = [];
  if (fontSize) properties.push(`font-size: ${fontSize}`);
  if (fontDecoration) properties.push(`font-decoration: ${fontDecoration}`);
  if (fontWeight) properties.push(`font-weight: ${fontWeight}`);

  return arrayToCSS(properties);
}

export function getSizeProperties({
  height,
  width,
  responsive,
  padding,
  margin
}) {
  const properties = [];
  if (width) {
    if (responsive) properties.push(`max-width: ${width}`, `width: 100%`);
    else properties.push(`width: ${width}`);
  }
  if (height) {
    if (responsive) properties.push(`max-height: ${height}`, `height: 100%`);
    else properties.push(`height: ${height}`);
  }
  return arrayToCSS(properties);
}

function returnDefault(property, type, defaultValue) {
  return typeof property === type ? property : defaultValue;
}

export function getLayoutProperties({ padding, margin, position }) {
  const properties = [];
  if (padding) properties.push(`padding: ${padding}`);
  if (margin) properties.push(`margin: ${margin}`);

  position = returnDefault(position, "string", "relative");
  if (position) properties.push(`position: ${position}`);
  return arrayToCSS(properties);
}

// TODO: Rename this function to something other than "Colouring"
export function getColouringProperties({
  background,
  backgroundColor,
  backgroundImage,
  color
}) {
  const properties = [];
  if (color) properties.push(`color: ${color}`);
  if (background) properties.push(`background: ${background}`);
  if (backgroundColor) properties.push(`background-color: ${backgroundColor}`);
  if (backgroundImage) properties.push(`background-image: ${backgroundImage}`);

  return arrayToCSS(properties);
}
