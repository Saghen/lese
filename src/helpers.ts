function arrayToCSS(array: string[]) {
  return `${array.join(";")};`;
}

function returnDefault(property: any, defaultValue: any) {
  return typeof property === "string" ? property : defaultValue;
}

function camelToKebab(string: string) {
  return string.replace(/([a-z0-9]|(?=[A-Z]))([A-Z])/g, "$1-$2").toLowerCase();
}

export type PropertyGeneratorHandler<T> = (props: T) => string;

export interface PropertyGeneratorOptions<T> {
  default?: string;
  handler?: PropertyGeneratorHandler<T>;
  property?: string;
}

export type PropertyGeneratorKey<T> = Array<[Extract<keyof T, string>, PropertyGeneratorOptions<T> | PropertyGeneratorHandler<T>] | Extract<keyof T, string>>;

export function propertyGenerator<T>(keys: PropertyGeneratorKey<T>): (props: T) => string {
  return (props: T) => {
    const properties = [];
    for (const key of keys) {
      // Exits
      if (Array.isArray(key) && !props[key[0]]) continue;
      if (typeof key === "string" && !props[key]) continue;

      if (!Array.isArray(key)) {
        props[key] && properties.push(`${camelToKebab(key)}: ${props[key]}`);
        continue;
      }

      const isFunction = typeof key[1] === "function";
      const isObject = typeof key[1] === "object";

      if (!(isFunction || isObject))
        throw new TypeError(`Invalid options provided at key: ${key[0]}`);

      // All validated
      if (isFunction) properties.push((key[1] as PropertyGeneratorHandler<T>)(props));

      if (isObject) {
        const options = key[1] as PropertyGeneratorOptions<T>;
        const property = options.property || camelToKebab(key[0]);
        if (options.default) props[key[0]] = returnDefault(props[key[0]], options.default);

        if (typeof options.handler === "function") properties.push(options.handler(props));
        else properties.push(`${property}: ${props[key[0]]}`);
      }
    }
    return arrayToCSS(properties);
  };
}
