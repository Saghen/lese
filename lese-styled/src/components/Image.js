import styled from "styled-components";
import { propertyGenerator } from "../helpers";

// TODO: Implement width and height
const imageProperties = propertyGenerator([
  ["responsive", () => "max-width: 100%;"],
  ["center", () => "object-position: center;"],
  ["cover", () => "object-fit: cover"]
]);

export default styled.img`
  ${imageProperties}
`;
