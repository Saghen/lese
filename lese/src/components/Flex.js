import styled from "@emotion/styled";
import Base from "./Base";
import { Children } from "react";

import { propertyGenerator } from "../helpers";

const getFlexProperties = propertyGenerator([
  ["column", () => "flex-direction: column"],
  [
    "xAlign",
    {
      default: "center",
      handler: ({ column, xAlign }) =>
        column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`,
    },
  ],
  [
    "yAlign",
    {
      default: "center",
      handler: ({ column, yAlign }) =>
        column ? `justify-content: ${yAlign}` : `align-items: ${yAlign}`,
    },
  ],
  [
    "noMarginReset",
    () => `> * {
       margin: 0;
      }`,
  ],
  [
    "separation",
    ({ column, separation }) => `
    > * + * {
      ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`}
    }
    `,
  ],

  ["wrap", { default: "wrap", property: "flex-wrap" }],
]);

const getChildFlexProperties = propertyGenerator([
  [
    "xAlignSelf",
    {
      default: "center",
      handler: ({ column, xAlignSelf }) =>
        column ? `align-self: ${xAlignSelf}` : `justify-self: ${xAlignSelf}`,
    },
  ],
  [
    "yAlignSelf",
    {
      default: "center",
      handler: ({ column, yAlignSelf }) =>
        column ? `justify-self: ${yAlignSelf}` : `align-self: ${yAlignSelf}`,
    },
  ],
]);

export default styled(Base)`
  display: flex;
  ${getFlexProperties}
  ${({ children, column }) => {
    const childrenArray = Children.toArray(children).filter(
      (elem) => typeof elem !== "string"
    );

    const properties = [];
    for (const [i, { props }] of childrenArray.entries())
      if (props && (props.xAlignSelf || props.yAlignSelf))
        properties.push(`
        > *:nth-child(${i + 1}) /* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */ {
          ${getChildFlexProperties({ ...props, column })}
        }
        `);
    return properties;
  }}
`;
