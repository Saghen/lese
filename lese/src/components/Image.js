import styled from "@emotion/styled";
import { propertyGenerator } from "../helpers";

// TODO: Implement width and height
const imageProperties = propertyGenerator([
  ["responsive", () => "max-width: 100%;"],
  ["center", () => "object-position: center;"],
  ["cover", () => "object-fit: cover"]
]);

export default styled.img`
  ${imageProperties}
`;
