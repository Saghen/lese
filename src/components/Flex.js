import { css } from "@emotion/core";
import styled from "@emotion/styled";
import Base from "./Base";

import { propertyGenerator } from "../helpers";
import { CLIENT_RENEG_LIMIT } from "tls";

const getFlexProperties = propertyGenerator([
  ["column", () => "flex-direction: column"],
  [
    "xAlign",
    {
      default: "center",
      handler: ({ column, xAlign }) =>
        column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`
    }
  ],
  [
    "yAlign",
    {
      default: "center",
      handler: ({ column, yAlign }) =>
        column ? `justify-content: ${yAlign}` : `align-items: ${yAlign}`
    }
  ],
  [
    ({ noMarginReset }) =>
      !noMarginReset &&
      `> * {
       margin: 0;
      }`
  ],
  [
    "separation",
    ({ column, separation }) => `
    > * + * {
      ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`}
    }
    `
  ],

  ["wrap", { default: "wrap", property: "flex-wrap" }]
]);

const getChildFlexProperties = propertyGenerator([
  [
    "xAlignSelf",
    {
      default: "center",
      handler: ({ column, xAlignSelf }) =>
        column ? `align-self: ${xAlignSelf}` : `justify-self: ${xAlignSelf}`
    }
  ],
  [
    "yAlignSelf",
    {
      default: "center",
      handler: ({ column, yAlignSelf }) =>
        column ? `justify-self: ${yAlignSelf}` : `align-self: ${yAlignSelf}`
    }
  ]
]);

export default styled(Base)`
  display: flex;
  ${getFlexProperties}
  ${({ children, column }) => {
    if (!Array.isArray(children)) return;
    const properties = [];
    for (const [i, { props }] of children.entries())
      if (props.xAlignSelf || props.yAlignSelf)
        properties.push(`
        > *:nth-child(${i + 1}) {
          ${getChildFlexProperties({ ...props, column })}
        }
        `);
    return properties;
  }}
`;
