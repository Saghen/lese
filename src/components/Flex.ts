import React from "react";
import styled from "@emotion/styled";
import Base from "./Base";
import { propertyGenerator } from "../helpers";

export interface FlexProps {
  column?: boolean;
  reverse?: boolean;
  align?: boolean;
  xAlign?: boolean | string;
  yAlign?: boolean | string;
  separation?: string | string[];
  wrap?: boolean | string;
}

const ignoreFlag =
  "/* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */";

const separationHandler = ({ column, separation }: FlexProps) => {
  const isString = typeof separation === "string";
  const isStringArray = Array.isArray(separation);
  if (!isString && !isStringArray) {
    throw new Error("separation must be a string or string array");
  }

  const getStyleBlock = (selector: string, separation: string) => `
      > ${selector} ${ignoreFlag} {
        ${column ? `margin-top: ${separation}` : `margin-left: ${separation}`};
        ${ignoreFlag}
      }
    `;

  const separations = isStringArray ? separation : separation.split(" ").filter(Boolean);
  if (separations.length > 1 || isStringArray) {
    const isRestSeparation = (separation: string) => separation.startsWith("...");
    const restSeparations = separations.filter(isRestSeparation);
    if (restSeparations.length > 1) {
      throw new Error("separation can only contain one rest operator");
    }

    const restSeparationIndex = separations.findIndex((a) => isRestSeparation(a));
    const startingSeparations = separations.slice(
      0,
      restSeparationIndex === -1 ? separations.length : restSeparationIndex
    );
    const endingSeparations = separations.slice(
      restSeparationIndex === -1 ? separations.length : restSeparationIndex + 1
    );

    return [
      ...restSeparations.map((separation) => getStyleBlock("* + *", separation.slice(3))),
      // Order is important since we want startingSeparations to replace endingSeparations
      ...endingSeparations.map((separation, i) =>
        getStyleBlock(`*:nth-last-child(${endingSeparations.length - i})`, separation)
      ),
      ...startingSeparations.map((separation, i) =>
        getStyleBlock(`*:first-child ${new Array(i + 1).fill("+ *").join(" ")}`, separation)
      ),
    ].join("\n");
  }
  return getStyleBlock("* + *", separation);
};

const applyDefaultProps =
  <T>(defaultProps: Partial<T>, callback: (arg: T) => any) =>
  (props: T) =>
    callback({ ...defaultProps, ...props });

const getFlexProperties = propertyGenerator<FlexProps>([
  ["column", () => `flex-direction: column`],
  ["reverse", { handler: ({ column }) => `flex-direction: ${column ? "column" : "row"}-reverse` }],
  [
    "align",
    {
      default: "center",
      handler: ({ align }) => `align-items: ${align}; justify-content: ${align}`,
    },
  ],
  [
    "xAlign",
    {
      default: "center",
      handler: ({ column, xAlign }) => (column ? `align-items: ${xAlign}` : `justify-content: ${xAlign}`),
    },
  ],
  [
    "yAlign",
    {
      default: "center",
      handler: ({ column, yAlign }) => (column ? `justify-content: ${yAlign}` : `align-items: ${yAlign}`),
    },
  ],
  ["separation", separationHandler],
  ["wrap", { default: "wrap", property: "flex-wrap" }],
]);

const getRowProperties = applyDefaultProps({ column: false }, getFlexProperties);
const getColumnProperties = applyDefaultProps({ column: true }, getFlexProperties);

export default styled(Base)<React.PropsWithChildren<FlexProps>>`
  display: flex;
  ${getFlexProperties}
`;

export const Row = styled(Base)<React.PropsWithChildren<FlexProps>>`
  display: flex;
  ${getRowProperties}
`;

export const Column = styled(Base)<React.PropsWithChildren<FlexProps>>`
  display: flex;
  ${getColumnProperties}
`;
