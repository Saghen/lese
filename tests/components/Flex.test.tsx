import React from "react";
import { Column, Row } from "../../src/components/Flex";
import { matchers } from "@emotion/jest";
import { renderComponent } from "./helpers";

expect.extend(matchers);

it("should use default values when passed true", () => {
  expect(renderComponent(<Row reverse />)).toHaveStyleRule("flex-direction", "row-reverse");
  expect(renderComponent(<Row xAlign />)).toHaveStyleRule("justify-content", "center");
  expect(renderComponent(<Row yAlign />)).toHaveStyleRule("align-items", "center");
  expect(renderComponent(<Row wrap />)).toHaveStyleRule("flex-wrap", "wrap");

  expect(renderComponent(<Column reverse />)).toHaveStyleRule("flex-direction", "column-reverse");
  expect(renderComponent(<Column xAlign />)).toHaveStyleRule("align-items", "center");
  expect(renderComponent(<Column yAlign />)).toHaveStyleRule("justify-content", "center");
  expect(renderComponent(<Column wrap />)).toHaveStyleRule("flex-wrap", "wrap");
});

it("should use values passed as props", () => {
  expect(renderComponent(<Row xAlign="flex-start" />)).toHaveStyleRule("justify-content", "flex-start");
  expect(renderComponent(<Row yAlign="flex-end" />)).toHaveStyleRule("align-items", "flex-end");
  expect(renderComponent(<Row wrap="wrap-reverse" />)).toHaveStyleRule("flex-wrap", "wrap-reverse");

  expect(renderComponent(<Column xAlign="flex-start" />)).toHaveStyleRule("align-items", "flex-start");
  expect(renderComponent(<Column yAlign="flex-end" />)).toHaveStyleRule("justify-content", "flex-end");
  expect(renderComponent(<Column wrap="wrap-reverse" />)).toHaveStyleRule("flex-wrap", "wrap-reverse");
});

it("should separate items via the separation prop", () => {
  expect(renderComponent(<Row separation="8px" />)).toMatchSnapshot();
  expect(renderComponent(<Row separation="8px 12px ...16px" />)).toMatchSnapshot();
  expect(renderComponent(<Row separation="...16px" />)).toMatchSnapshot();
  expect(renderComponent(<Row separation="8px ...16px" />)).toMatchSnapshot();
  expect(renderComponent(<Row separation="...16px 8px" />)).toMatchSnapshot();
  expect(renderComponent(<Row separation="8px ...16px 12px" />)).toMatchSnapshot();
  expect(renderComponent(<Row separation="8px 10px ...16px 12px 14px" />)).toMatchSnapshot();

  expect(renderComponent(<Column separation="8px" />)).toMatchSnapshot();
  expect(renderComponent(<Column separation="8px 12px ...16px" />)).toMatchSnapshot();
  expect(renderComponent(<Column separation="...16px" />)).toMatchSnapshot();
  expect(renderComponent(<Column separation="8px ...16px" />)).toMatchSnapshot();
  expect(renderComponent(<Column separation="...16px 8px" />)).toMatchSnapshot();
  expect(renderComponent(<Column separation="8px ...16px 12px" />)).toMatchSnapshot();
  expect(renderComponent(<Column separation="8px 10px ...16px 12px 14px" />)).toMatchSnapshot();
});
