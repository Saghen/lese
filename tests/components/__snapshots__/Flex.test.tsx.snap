// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should separate items via the separation prop 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 2`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: 16px;
}

.emotion-0>*:first-child {
  margin-left: 8px;
}

.emotion-0>*:first-child+* {
  margin-left: 12px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 3`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: ...16px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 4`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: 16px;
}

.emotion-0>*:first-child {
  margin-left: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 5`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: 16px;
}

.emotion-0>*:nth-last-child(1) {
  margin-left: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 6`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: 16px;
}

.emotion-0>*:nth-last-child(1) {
  margin-left: 12px;
}

.emotion-0>*:first-child {
  margin-left: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 7`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-0>*+* {
  margin-left: 16px;
}

.emotion-0>*:nth-last-child(2) {
  margin-left: 12px;
}

.emotion-0>*:nth-last-child(1) {
  margin-left: 14px;
}

.emotion-0>*:first-child {
  margin-left: 8px;
}

.emotion-0>*:first-child+* {
  margin-left: 10px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 8`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 9`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: 16px;
}

.emotion-0>*:first-child {
  margin-top: 8px;
}

.emotion-0>*:first-child+* {
  margin-top: 12px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 10`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: ...16px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 11`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: 16px;
}

.emotion-0>*:first-child {
  margin-top: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 12`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: 16px;
}

.emotion-0>*:nth-last-child(1) {
  margin-top: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 13`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: 16px;
}

.emotion-0>*:nth-last-child(1) {
  margin-top: 12px;
}

.emotion-0>*:first-child {
  margin-top: 8px;
}

<div
  className="emotion-0"
/>
`;

exports[`should separate items via the separation prop 14`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0>*+* {
  margin-top: 16px;
}

.emotion-0>*:nth-last-child(2) {
  margin-top: 12px;
}

.emotion-0>*:nth-last-child(1) {
  margin-top: 14px;
}

.emotion-0>*:first-child {
  margin-top: 8px;
}

.emotion-0>*:first-child+* {
  margin-top: 10px;
}

<div
  className="emotion-0"
/>
`;
